src/Umbraco.Web/Routing/RedirectTrackingEventHandler.cs

```diff
@@ -89,62 +90,92 @@ protected override void ApplicationStarted(UmbracoApplicationBase umbracoApplica
         {
-            get { return Moving && UmbracoContext.Current.HttpContext.Items[ContextKey2] != null; }
+            get
+            {
+                return Moving && RequestCache.GetCacheItem(ContextKey2) != null;
+            }
             set
             {
                 if (Moving && value)
-                    UmbracoContext.Current.HttpContext.Items[ContextKey2] = true;
+                {
+                    //this forces true into the cache
+                    RequestCache.GetCacheItem(ContextKey2, () => true);
+                }
                 else
-                    UmbracoContext.Current.HttpContext.Items.Remove(ContextKey2);
+                {
+                    RequestCache.ClearCacheItem(ContextKey2);
+                }
             }
         }
 
         private static bool Moving
         {
-            get { return UmbracoContext.Current.HttpContext.Items[ContextKey1] != null; }
+            get { return RequestCache.GetCacheItem(ContextKey1) != null; }
             set
             {
                 if (value)
-                    UmbracoContext.Current.HttpContext.Items[ContextKey1] = true;
+                {
+                    //this forces true into the cache
+                    RequestCache.GetCacheItem(ContextKey1, () => true);
+                }
                 else
                 {
-                    UmbracoContext.Current.HttpContext.Items.Remove(ContextKey1);
-                    UmbracoContext.Current.HttpContext.Items.Remove(ContextKey2);
+                    RequestCache.ClearCacheItem(ContextKey1);
+                    RequestCache.ClearCacheItem(ContextKey2);
                 }
             }
         }
 
+        /// <summary>
+        /// Before the items are published, we need to get it's current URL before it changes
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="args"></param>
         private static void ContentService_Publishing(IPublishingStrategy sender, PublishEventArgs<IContent> args)
         {
             if (LockedEvents) return;
 
-            var contentCache = UmbracoContext.Current.ContentCache;
+            var contentCache = GetPublishedCache();
+            if (contentCache == null) return;
+
             foreach (var entity in args.PublishedEntities)
             {
-                var entityContent = contentCache.GetById(entity.Id);
-                if (entityContent == null) continue;
-                foreach (var x in entityContent.DescendantsOrSelf())
+                //don't continue if this entity hasn't changed with regards to anything 
+                // that might change it's URLs
+                if (entity.IsDirty() == false) continue;
+
+                if (entity.IsPropertyDirty("Name") 
+                    || entity.IsPropertyDirty(Constants.Conventions.Content.UrlName)
+                    || entity.IsPropertyDirty(Constants.Conventions.Content.UrlAlias)
+                    || Moving)
```

Comment:
We don't want to iterate all descendents on all publishing - this will destroy performance. So the dirty check is one thing but are you saying I should call in to the IUrlSegmentProvider for this item at this stage to store what it is, then re-check it after publishing ? ... or let me know what exactly you think needs to be done.